### PR TITLE
PDCT 251/improve lang

### PR DESCRIPTION
**Pull request by diversemix**
# Description

- Adds new flags to the PhysicalDocumentLanguage
- Adds db migrations
- Updates tests for pipeline to use the `source` = `MODEL`
- Updates ingestion to use `source` = `USER`
- Renames `_BaseModelEnum`  to `_BaseModelEnum` moved into new file to avoid circular imports

## Type of change

Please select the option(s) below that are most relevant:

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change

## Test Result

The following was tested locally:
- start from blank database creates correct schema
- start from existing staging data - also creates correct schema with the defaults

## Reviewer Checklist

- [ ] The PR represents a single feature (small driveby fixes are also ok)
- [ ] The PR includes tests that are sufficient for the level of risk
- [ ] The code is sufficiently commented, particularly in hard-to-understand areas
- [ ] Any required documentation updates have been made
- [ ] Any TODOs added are captured in future tickets
- [ ] No FIXMEs remain

### CODE REVIEW

File: app/core/ingestion/family.py [hunk 1]
```diff
@@ -183,6 +186,9 @@ def none_if_empty(data: str) -> Optional[str]:
                 params.document_title,
                 family_document.physical_document,
             )
+            update_physical_document_languages(
+                db, params.language, result, family_document.physical_document
+            )
```
Comment (joel-wright) [inline app/core/ingestion/family.py:191]: Is deleting all existing languages on every ingest not a little heavy handed? I had wondered if we could label all existing languages as not visible with a one-off script, then this script method would only need to add a missing language or mark an existing one as "user + visible" if necessary... 

Also, this makes it impossible for the pipeline to add document languages unless we run a complete update each time (currently we only send updates).

File: app/db/crud/document.py [hunk 2]
```diff
@@ -102,21 +107,27 @@ def get_family_document_and_context(
...
+) -> Sequence[str]:
+    result = (
+        db.query(PhysicalDocumentLanguage.visible, Language.language_code)
+        .join(Language, Language.id == PhysicalDocumentLanguage.language_id)
+        .filter(PhysicalDocumentLanguage.document_id == physical_document.id)
+        .all()
+    )
+    if result is None or len(result) == 0:
+        return []
+    return [cast(str, code) for visible, code in result if visible]
```
[the remark this answers is not in the document]
Comment (diversemix) [reply app/db/crud/document.py:130]: I did think about this - but let's chat about it, as its not a standard "relationship" as it requires a filter - so ends up being a property/function anyhow

File: app/core/ingestion/family.py (hunk 1 above)
Comment (diversemix) [reply app/core/ingestion/family.py:191]: I don't believe the code does delete all existing languages on every ingest - perhaps you can point out where you think this is occurring?

File: app/core/ingestion/physical_document.py [hunk 3]
```diff
@@ -32,21 +36,53 @@ def create_physical_document_from_params(
...
-            result["physical_document_language"] = phys_doc_languages
+        doc_languages = result.get("language", [])
+        doc_languages.append(to_dict(lang))
+        result["language"] = doc_languages
 
-    return physical_document
+        physical_document_language = PhysicalDocumentLanguage(
+            language_id=lang.id,
+            document_id=physical_document.id,
+            source=LanguageSource.USER,
```
Comment (THOR300) [inline app/core/ingestion/physical_document.py:80]: I'm just wondering whether this function would be utilised by the db writer when calling back to the backend with identified languages and thus whether defaulting to a USER defined language is correct.

File: alembic/versions/0018_add_more_language_fields.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    #
+    # Code taken from: https://stackoverflow.com/a/65173731
+    language_source = postgresql.ENUM(LanguageSource, name="languagesource")
+    language_source.create(op.get_bind(), checkfirst=True)
+
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('physical_document_language', sa.Column(
+        'source',
+        language_source,
+        server_default=(LanguageSource.MODEL.upper()),
```
Comment (THOR300) [inline alembic/versions/0018_add_more_language_fields.py:33]: Just to confirm for all the physical documents we currently have in the database we would be assigning a default value of MODEL defined language?

File: app/core/ingestion/physical_document.py (hunk 3 above)
Comment (diversemix) [reply app/core/ingestion/physical_document.py:80]: Nope this is ingestion only - i.e. via CSV ... i.e. via a user :smile:

File: alembic/versions/0018_add_more_language_fields.py (hunk 4 above)
Comment (diversemix) [reply alembic/versions/0018_add_more_language_fields.py:33]: Yup - and also `visible` = False

**Review by joel-wright — APPROVED** [review]

**Pull request merged**